src/System.Private.CoreLib/shared/System/Globalization/DateTimeParse.cs

```diff
@@ -4831,6 +4838,215 @@ private static bool ParseFormatR(ReadOnlySpan<char> source, ref ParsingInfo pars
+                    return false;
+                }
+
+                year = (int)(y1*1000 + y2*100 + y3*10 + y4);
+            }
+
+            int month;
+            {
+                uint m1 = (uint)(source[5] - '0'), m2 = (uint)(source[6] - '0');
+
+                if (m1 > 9 || m2 > 9)
+                {
+                    result.SetBadDateTimeFailure();
+                    return false;
+                }
+
+                month = (int)(m1*10 + m2);
+            }
+
+            int day;
+            {
+                uint d1 = (uint)(source[8] - '0'), d2 = (uint)(source[9] - '0');
+
+                if (d1 > 9 || d2 > 9)
+                {
+                    result.SetBadDateTimeFailure();
+                    return false;
+                }
+
+                day = (int)(d1*10 + d2);
+            }
+
+            int hour;
+            {
+                uint h1 = (uint)(source[11] - '0'), h2 = (uint)(source[12] - '0');
+
+                if (h1 > 9 || h2 > 9)
+                {
+                    result.SetBadDateTimeFailure();
+                    return false;
+                }
+
+                hour = (int)(h1*10 + h2);
+            }
+
+            int minute;
+            {
+                uint m1 = (uint)(source[14] - '0'), m2 = (uint)(source[15] - '0');
+
+                if (m1 > 9 || m2 > 9)
+                {
+                    result.SetBadDateTimeFailure();
+                    return false;
+                }
+
+                minute = (int)(m1*10 + m2);
+            }
+
+            int second;
+            {
+                uint s1 = (uint)(source[17] - '0'), s2 = (uint)(source[18] - '0');
+
+                if (s1 > 9 || s2 > 9)
+                {
+                    result.SetBadDateTimeFailure();
+                    return false;
+                }
+
+                second = (int)(s1*10 + s2);
+            }
+
+            double fraction;
+            {
+                uint f1 = (uint)(source[20] - '0');
+                uint f2 = (uint)(source[21] - '0');
+                uint f3 = (uint)(source[22] - '0');
+                uint f4 = (uint)(source[23] - '0');
+                uint f5 = (uint)(source[24] - '0');
+                uint f6 = (uint)(source[25] - '0');
+                uint f7 = (uint)(source[26] - '0');
+
+                if (f1 > 9 || f2 > 9 || f3 > 9 || f4 > 9 || f5 > 9 || f6 > 9 || f7 > 9)
+                {
+                    result.SetBadDateTimeFailure();
+                    return false;
+                }
+
+                fraction = (f1*1000000 + f2*100000 + f3*10000 + f4*1000 + f5*100 + f6*10 + f7) / 10000000.0;
+            }
+
+            if (!DateTime.TryCreate(year, month, day, hour, minute, second, 0, out DateTime dateTime))
+            {
+                result.SetBadDateTimeFailure();
+                return false;
+            }
+            result.parsedDate = dateTime.AddTicks((long)Math.Round(fraction * Calendar.TicksPerSecond));
```

Comment:
>dateTime.AddTicks [](start = 32, length = 17)

is This possible to throw and we get different exception than what we used to get before?

Reply:
By construction I don't believe that's possible, as DateTime.MaxValue is "9999-12-31T23:59:59.9999999", and by this point we've already validated that the DateTime of ""9999-12-31T23:59:59" was created successfully and validated against the calendar.  We've also validated that none of the fraction digits are greater than 9, and "9999999" is the largest possible value this can be.

Is there another way this could manifest?  We were very skimpy on coverage here, so I added a bunch of tests in https://github.com/dotnet/corefx/pull/30860, including for MaxValue:
https://github.com/dotnet/corefx/pull/30860/files#diff-a7a9c52518e52358d4f3ab8e88cbf53fR963
but if you see something missing, I can rectify that.

Reply:
"9999-12-31T23:59:59.9999999Z" if you parse this when having TZ before UTC (e.g. Fuji +12) this should fail the parsing. in most cases, we'll get the right exception but there is will be specific values which can cause the exception only when adding the fraction ticks per second. I can try to get the exact data if you like.

Reply:
> this should fail the parsing

Yes... but the factoring in of timezone doesn't happen until after this point, down at the bottom of the method where it calls DetermineTimeZoneAdjustments.  At this point, we've called DateTime.TryCreate, using only the data from "9999-12-31T23:59:59", creating a DateTimeKind.Unspecified, and we add the "9999999" fraction to that.  It's only after this AddTicks that we then parse the 'Z' and factor in the timezone as part of DetermineTimeZoneAdjustments.

Reply:
ok, we should be fine then. thanks for clarifying.

Reply:
Cool, thanks.

Reply:
It makes me happy that 7,981 years from now (perhaps on another star system) our code will function correctly because we took care to correctly handle the year 9,999. 😃

Reply:
Yes, we're very forward thinking.